core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatEntry.scala

```diff
@@ -399,7 +398,7 @@ final class ColumnFormatValue
   override def toData(out: DataOutput): Unit = {
     val buffer = getBufferRetain
     try {
-      val numBytes = buffer.remaining()
+      val numBytes = buffer.limit()
```

Comment:
Right internal buffer is never touched once set so its position/limit is fixed. The getBufferRetain either returns a duplicate or slice to ensure that returned buffer always has initial position as 0.